install/kubernetes/helm/istio/charts/ingressgateway/templates/deployment.yaml

```diff
@@ -28,9 +28,9 @@ spec:
           image: "{{ .Values.global.hub }}/proxyv2:{{ .Values.global.tag }}"
           imagePullPolicy: {{ .Values.global.imagePullPolicy }}
           ports:
-            {{- range $key, $val := .Values.deployment.ports }}
-            - containerPort: {{ $val.containerPort }}
-            {{- end }}
+          - containerPort: 80
+          - containerPort: 443
+          - containerPort: 31400
```

Comment:
I believe @rshriram 's intent here was to permit the ability to add any port for ingressgateway.  @rshriram can you confirm?

Reply:
Yes, this is not correct. I wanted the ability for people to expose custom ports other than 80 and 443. So, lets keep this as it is.

Reply:
@sdake @rshriram I think the logic here should be as follows:

```
ports:
- containerPort: 80
- containerPort: 443
- containerPort: 31400
{{- range $key, $val := .Values.deployment.ports }}
- containerPort: {{ $val.containerPort }}
{{- end }}
```

Reply:
@gyliu513 if the logic is wrong lets fix that problem.  However, ports: should still go in values file per @rshriram's comments.

Reply:
@rshriram @sdake Just one point I don't understand, should the container ports be fixed for each container, why we expose that to user?

Reply:
@sdake @rshriram 

I think it is not good to expose 80, 443 and 31400 in `values.yaml`. If customer delete those port by mistake, then ingressgateway will not work.

So the proposal here is use following logic: always contain 80, 443 and 31400, and then handle customized ports.

```
ports:
- containerPort: 80
- containerPort: 443
- containerPort: 31400
{{- range $key, $val := .Values.deployment.ports }}
- containerPort: {{ $val.containerPort }}
{{- end }}
```

One more question to @rshriram , for which case does customer want to expose more ports? And Why do they want to expose more ports? How does the customer know which ports does the proxyv2 image which is used by ingressgateway can expose?

Reply:
@rshriram @sdake comments for this? IMHO, we should not expose those ports to customize as those ports are controlled by Istio but not user.

Reply:
ITs just a bunch of ports for which they want to configure the load balancer. So, someone may choose to not even use 80/443, but have 9080 and 9443. The process in this container is an Envoy and it will only listen on the ports defined in the Gateway.yaml, which is authored by the end user. So, lets please not hard code ports in these yamls. I've found that people often do not know where to tweak the values. Having them all in one single values.yaml is of great help (without any hardcoded defaults that might actually be incorrect).

Reply:
@rshriram  So you mean I can define any port but not 80/443, as long as I also put those ports to Gateway and it will work.

One example is that I can define port as follows:
```
 ports:
- containerPort: 8080
```

And Gateway define as 
```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: bookinfo-gateway
spec:
  selector:
    istio: ingressgateway # use istio default controller
  servers:
  - port:
      number: 8080
      name: http
      protocol: HTTP
    hosts:
    - "*"
```